<a href="https://colab.research.google.com/github/hrulevskaya/ML_tasks/blob/main/%D0%9F%D1%80%D0%BE%D0%B3%D0%BD%D0%BE%D0%B7%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D1%86%D0%B5%D0%BD%D1%8B%20%D0%BF%D0%BE%D0%B4%D0%B5%D1%80%D0%B6%D0%B0%D0%BD%D0%BD%D1%8B%D1%85%20%D0%B0%D0%B2%D1%82%D0%BE%D0%BC%D0%BE%D0%B1%D0%B8%D0%BB%D0%B5%D0%B9%20(%D0%BB%D0%B8%D0%BD%D0%B5%D0%B9%D0%BD%D0%B0%D1%8F%20%D1%80%D0%B5%D0%B3%D1%80%D0%B5%D1%81%D1%81%D0%B8%D1%8F%20/%20%D0%BB%D0%B0%D1%81%D1%81%D0%BE%20/%20%D0%B3%D1%80%D0%B5%D0%B1%D0%BD%D0%B5%D0%B2%D0%B0%D1%8F%20%D1%80%D0%B5%D0%B3%D1%80%D0%B5%D1%81%D1%81%D0%B8%D1%8F).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/used_cars.csv")

columns = ["brand", "model_year", "milage", "fuel_type", "transmission", "accident", "clean_title", "price"]

data = data[columns]
data

,brand,model_year,milage,fuel_type,transmission,accident,clean_title,price
0,Ford,2013,"51,000 mi.",E85 Flex Fuel,6-Speed A/T,At least 1 accident or damage reported,Yes,"$10,300"
1,Hyundai,2021,"34,742 mi.",Gasoline,8-Speed Automatic,At least 1 accident or damage reported,Yes,"$38,005"
2,Lexus,2022,"22,372 mi.",Gasoline,Automatic,None reported,NaN,"$54,598"
3,INFINITI,2015,"88,900 mi.",Hybrid,7-Speed A/T,None reported,Yes,"$15,500"
4,Audi,2021,"9,835 mi.",Gasoline,8-Speed Automatic,None reported,NaN,"$34,999"
...,...,...,...,...,...,...,...,...
4004,Bentley,2023,714 mi.,Gasoline,8-Speed Automatic with Auto-Shift,None reported,Yes,"$349,950"
4005,Audi,2022,"10,900 mi.",Gasoline,Transmission w/Dual Shift Mode,None reported,Yes,"$53,900"
4006,Porsche,2022,"2,116 mi.",NaN,Automatic,None reported,NaN,"$90,998"
4007,Ford,2020,"33,000 mi.",Gasoline,A/T,None reported,Yes,"$62,999"


In [ ]:
from collections import Counter
data["clean_title"] = data["clean_title"].replace({np.nan: False, "Yes": True})
filtered_data = data.dropna()
cars_per_brand = Counter(data["brand"])
keep_brands = [k for k, v in cars_per_brand.items() if v >= 20]
filtered_data = filtered_data[filtered_data['brand'].isin(keep_brands)]



In [ ]:
filtered_data["milage"] = filtered_data["milage"].apply(lambda x: int(x[:-4].replace(",", "")))
filtered_data["price"] = filtered_data["price"].apply(lambda x: int(x[1:].replace(",", "")))

non_numeric_columns = filtered_data.select_dtypes(exclude=['number']).columns
filtered_data = pd.get_dummies(filtered_data, columns = non_numeric_columns, drop_first=True)
filtered_data

,model_year,milage,price,brand_Audi,brand_BMW,brand_Bentley,brand_Buick,brand_Cadillac,brand_Chevrolet,brand_Chrysler,...,transmission_M/T,transmission_Manual,"transmission_Manual, 6-Spd",transmission_SCHEDULED FOR OR IN PRODUCTION,transmission_Transmission Overdrive Switch,transmission_Transmission w/Dual Shift Mode,transmission_Variable,transmission_–,accident_None reported,clean_title_True
0,2013,51000,10300,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,True
1,2021,34742,38005,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,True
2,2022,22372,54598,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False
3,2015,88900,15500,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,True
4,2021,9835,34999,True,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4003,2018,53705,25900,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,True
4004,2023,714,349950,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,True,True
4005,2022,10900,53900,True,False,False,False,False,False,False,...,False,False,False,False,False,True,False,False,True,True
4007,2020,33000,62999,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,True


In [ ]:
filtered_data.describe()

,model_year,milage,price
count,3619.000000,3619.000000,3.619000e+03
mean,2015.391821,67159.373308,4.182714e+04
std,6.050332,52646.270628,7.103754e+04
min,1974.000000,100.000000,2.000000e+03
25%,2012.000000,25006.500000,1.690000e+04
50%,2017.000000,56000.000000,3.000000e+04
75%,2020.000000,96409.500000,4.800000e+04
max,2024.000000,405000.000000,2.954083e+06


In [ ]:
import numpy as np
import pandas as pd
import scipy.io as io
from sklearn import metrics
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression, Lasso, Ridge
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

In [ ]:
X = filtered_data.drop(columns=['price'])
y = filtered_data['price']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
def train_model(model):
  model.fit(X_train_scaled, y_train)
  print("=" * 30)
  print(model)
  y_pred_scaled = model.predict(X_train_scaled)
  rmse_scaled = mean_squared_error(y_train, y_pred_scaled) ** 0.5
  mae_scaled = mean_absolute_error(y_train, y_pred_scaled)
  r2_scaled = r2_score(y_train, y_pred_scaled)
  print("train:")
  print(f"\tRMSE: {rmse_scaled}")
  print(f"\tMAE: {mae_scaled}")
  print(f"\tR2: {r2_scaled}")
  y_pred_scaled = model.predict(X_test_scaled)
  rmse_scaled = mean_squared_error(y_test, y_pred_scaled) ** 0.5
  mae_scaled = mean_absolute_error(y_test, y_pred_scaled)
  r2_scaled = r2_score(y_test, y_pred_scaled)
  print()
  print("test:")
  print(f"\tRMSE: {rmse_scaled}")
  print(f"\tMAE: {mae_scaled}")
  print(f"\tR2: {r2_scaled}")
  print("=" * 30)
  print()
  return model

In [ ]:
train_model(LinearRegression())
train_model(Lasso(1))
train_model(Ridge(1))

LinearRegression()
train:
	RMSE: 65838.17724080422
	MAE: 18101.943225336254
	R2: 0.2521712501522855

test:
	RMSE: 34679.53960537845
	MAE: 18938.254807824942
	R2: 0.4105257835381205

Lasso(alpha=1)
train:
	RMSE: 65838.30975250961
	MAE: 18103.14551998181
	R2: 0.2521682398574425

test:
	RMSE: 34702.37739634007
	MAE: 18915.879564164574
	R2: 0.4097491456819231

Ridge(alpha=1)
train:
	RMSE: 65838.27070906686
	MAE: 18100.446011717915
	R2: 0.2521691268157489

test:
	RMSE: 34693.21996093052
	MAE: 18917.67041102717
	R2: 0.4100606212237129



Ridge(alpha=1)

In [ ]:
param_grid = {'alpha': [0] + [10 ** x for x in range(-5, 6)]}
print(param_grid)

print("BEST MODELS")
print()
lasso_gs = GridSearchCV(Lasso(random_state=0), param_grid, scoring='neg_mean_squared_error', cv=3)
lasso_gs.fit(X_train_scaled, y_train)
best_alpha = lasso_gs.best_estimator_.alpha
lasso_model = train_model(Lasso(best_alpha))

ridge_gs = GridSearchCV(Ridge(random_state=0), param_grid, scoring='neg_mean_squared_error', cv=3)
ridge_gs.fit(X_train_scaled, y_train)
best_alpha = ridge_gs.best_estimator_.alpha
ridge_model = train_model(Ridge(best_alpha))


{'alpha': [0, 1e-05, 0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000, 100000]}
BEST MODELS

Lasso(alpha=1000)
train:
	RMSE: 66264.99571094982
	MAE: 18569.464761196155
	R2: 0.24244371041378243

test:
	RMSE: 34292.0872755109
	MAE: 18836.20125387043
	R2: 0.4236238416876903

Ridge(alpha=1000)
train:
	RMSE: 66386.17428855407
	MAE: 17677.115536492605
	R2: 0.23967049570374088

test:
	RMSE: 33252.41808502303
	MAE: 17661.03153447981
	R2: 0.45804324737630364



Данные обработаны: категориальные признаки закодированы (One-Hot), числовые — отмасштабированы (StandardScaler).

Базовая модель (LinearRegression): показала R² = 0.41 на тесте.

Регуляризация: Lasso и Ridge с параметрами по умолчанию (alpha=1) не дали прироста качества.

Подбор гиперпараметров (GridSearchCV): выявил, что моделям нужна крайне сильная регуляризация (alpha=1000).

Финальный результат: лучшей стала Ridge (alpha=1000) с R² = 0.458. Прирост качества по сравнению с базовой моделью составил около +5 п.п.